### ISSUE TITLE
Lost exception breakpoints for multi thread managed code.

### ISSUE THREAD

**Issue by a-zh** [OmniSharp/omnisharp-vscode#3369]
1. Open Visual Studio Code IDE 1.39.1
2. Bring up the command palette (press <kbd>F1</kbd>)

CPUs | Intel(R) Core(TM) i9-7900X CPU @ 3.30GHz (20 x 3312)
-- | --
GPU Status | 2d_canvas: enabledflash_3d: enabledflash_stage3d: enabledflash_stage3d_baseline: enabledgpu_compositing: enabledmultiple_raster_threads: enabled_onnative_gpu_memory_buffers: disabled_softwareoop_rasterization: disabled_offprotected_video_decode: unavailable_offrasterization: enabledskia_deferred_display_list: disabled_offskia_renderer: disabled_offsurface_synchronization: enabled_onvideo_decode: enabledviz_display_compositor: disabled_offwebgl: enabledwebgl2: enabled
Load (avg) | undefined
Memory (System) | 63.68GB (50.29GB free)
Process Argv |  
Screen Reader | no
VM | 0%

Extension | Author (truncated) | Version
-- | -- | --
csharp | ms- | 1.21.5
debug | web | 0.23.1

## Environment data
`dotnet --info` output:
VS Code version:

dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.2.103
 Commit:    8edbc2570a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.103\

Host (useful for support):
  Version: 2.2.1
  Commit:  878dd11e62

.NET Core SDKs installed:
  1.1.14 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.503 [C:\Program Files\dotnet\sdk]
  2.2.103 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

C# Extension version:
C# 1.21.5 plugin for Visual Studio Code

## Steps to reproduce
You should take BUG1.cs and create the project. All breakpoints and exception filters should be removed. For reproducing run debugging. After entry point say F5 again. We expect exception breakpoint but unfortunately we can not see exception window.
Also the problem can get more seriously if from multi threads throws many exceptions at the same time. Please, check it too.

## Expected  behavior
For user applications with thread and for single thread applications with exceptions we want to have the same behavior.

## Actual behavior
Not exception window for multi-threading code.

[debug_console.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3762304/debug_console.txt)
[BUG1.cs.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3762281/BUG1.cs.txt)
[bug1.csproj.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3762662/bug1.csproj.txt)
[tasks.json.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3762664/tasks.json.txt)
[launch.json.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3762666/launch.json.txt)

## Please give answers on the questions:
1) Could you explain how to should work "User-Unhandled Excepitons" filter? 
If I set this flag or disable this flag which behavior are expected? Please give me some code examples for this. Thank you.
As I can expect "User-Unhandled" works only with user exception types and allow to stop on handled (inside try block) exceptions. Is it right?

**Comment by a-zh** [comment]
Important remark. Exceptions filter should be empty.

![image](https://user-images.githubusercontent.com/39411904/67401030-347df100-f5b7-11e9-97ab-7f623b9f1396.png)

**Comment by a-zh** [comment]
@WardenGnaw

**Comment by gregg-miskelly** [comment]
@a-zh I believe you opened this issue to try and understand what "User-Unhandled Exceptions" means. Is that correct? If not, can you clarify what you are asking?

"User-Unhandled" is something that applies when you have "Just My Code" enabled (which is enabled by default). The idea is that there are situations where a framework (non-user code) might call back into your code, and when your code throws an exception which is caught by the framework, this might be an indication of a problem, so the debugger should break. An example of such a situation is a button click handler - WinForms will catch any exception that the handler throws, and let the app continue working. But if you are debugging that application you probably want to know when this happens. It is different from Unhandled exceptions in that with an Unhandled exception, nothing has caught the exception. So the debugger breaks to prevent the process from just tearing down. You can't disable stopping on unhandled exceptions. It will always stop since this is the last chance to inspect the process before it exits. In your example, I would expect the exception raised from 'TestSystemException0' to be unhandled.

**Comment by a-zh** [comment]
@gregg-miskelly
Dear Gregg, please let me exaplain of my goals step by step:

0) Sorry for mixing of questions and report of bug in one ticket.
I supposed that it can be faster for resolving.

1) Context of my questions around the works for Exception breakpoints supporting for
opensource C# debugger NetcoreDBG.
Project: https://github.com/samsung/netcoredbg 
Commit: https://github.com/Samsung/netcoredbg/commit/9e705af8df16dd5727e1396c5280d2c9f21e227d
The reason of questions - NetcoreDBG wants to be like VSDBG.

2) After single thread support are ready PR for multithread support and now still waiting approval.
My questions, because I see inconsistent behavior for MI and VSCODE protocols in VSDBG.
For complete this works we want to have details for "User-Unhandled" behavior.
Please, look at this example (in attachment 
[BUG4_user_inner.cs.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3783921/BUG4_user_inner.cs.txt)) For this file:
 - I see 3 events for MI and 1 for VSCODE protocols from vsdbg. 
Is it correct?
 - I suppose, that we should get 3 events in VSCODE IDE with enabled "User-Unhandled" filter.
Is it correct?

2a) How debugger can to know that the current callback from user code? Which label for user code? Is it right that only by exception class name we can say about it?

2b) Do you have open documentations and examples, tests for check "User-Unhandled" behavior on MI and VSCODE protocols?

3) Multithread supporting was hard for NetcoreDBG because FuncEval is asynchronous event. 
In particular, a new Exception events was interrupts the evaluation process and lock continue process.
After fixing we checked VSCODE IDE and founds problems in VSDBG too (see above).

**Comment by a-zh** [comment]
@gregg-miskelly Could you give me any feedback (see above)?

**Comment by gregg-miskelly** [comment]
Since your questions are really for your own debugger, you can really make whatever choices you would like, but here are a few notes --

* Move away from MI: If you are using MI to plug into Visual Studio / VS Code, I would suggest moving away from it in favor of the debug adapter protocol. From Version 15.3 on vsdbg does all debugging through the debug adapter protocol. The only reason vsdbg still supports MI is that we haven't taken the time to remove support. Similarly, Visual Studio's MI client, MIEngine, may stop supporting the MI extensions that we added for clrdbg. So I would suggest against it for that reason too. If your are using MI for supporting non-Visual Studio editors, then that is a different story. But 
* In your example in [BUG4_user_inner.cs.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3783921/BUG4_user_inner.cs.txt), I wouldn't expect any user-unhandled exception. There are many exceptions thrown, but they are all caught by user code until the final exception from line 22, which is 'unhanded' rather than 'user-unhanded'. Vsdbg, by default, will generate message events for when the exceptions are thrown (`Exception thrown: 'AppException' in BUG4_user_inner.dll`).
* You can decide your own heuristic if you like, by vsdbg considers a module user code if it is both unoptimized and it can find a .pdb file. In terms of deciding if an exception is user-unhandled, the type of the exception doesn't matter. Just where the exception is thrown, caught, and what frames are between those two things.
* We don't have any open source tests for vsdbg
* What was the problem you found with vsdbg?

**Comment by a-zh** [comment]
Gregg thanks a lot for you answers.
 
>> What was the problem you found with vsdbg?

vsdbg has a problem with delivering of exception event from thread to vscode ide.
Please, look at BUG1.cs.txt. vscode ide + vsdbg and with empty exception filter: https://user-images.githubusercontent.com/39411904/67401030-347df100-f5b7-11e9-97ab-7f623b9f1396.png
Please compare behavior without thread.

**Comment by gregg-miskelly** [comment]
Ah thanks. Yes, this is https://github.com/OmniSharp/omnisharp-vscode/issues/3388.

Any more questions?

**Comment by a-zh** [comment]
Gregg, are you really understood me correctly? I try to say: if throw exception from Main for single thread application than all is OK, but if exception from any other threads, than I can not see exception window?

**Comment by gregg-miskelly** [comment]
@a-zh here is what I am seeing if I launch with 'User-unhanded' turned off. Is this what you are seeing?

Screen capture is below. The debugger will stop, but it will not navigate to a source file (because only non-user code is on the stack) and it will not bring up exception details view (again, because VS Code only does this if there is a source file). The output window will show you the type of exception thrown. If you disable justMyCode via launch.json, you can see the stack, and use $exception to dig into the details.

![image](https://user-images.githubusercontent.com/11078489/68251200-77888d00-ffd7-11e9-82e1-3cf364411475.png)

**Comment by a-zh** [comment]
Yes, surely! 

1) Unfortunately, I can't reproduce exception Window with disabled `"justMyCode": false,` in my `launch.json`. This is my `launch.json`:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": ".NET Core Launch (console) with pipeline",
            "type": "coreclr",
            "request": "launch",
            "preLaunchTask": "build",               
            "program": "${workspaceRoot}/bin/Debug/netcoreapp2.0/ClassLibrary1.dll",
            "args": [],
            "cwd": "${workspaceFolder}",
            "stopAtEntry": false,
            "console": "internalConsole",
            "internalConsoleOptions": "openOnSessionStart",
            "justMyCode": false,
            "logging": {
                "engineLogging": true
            },
        },
        {
            "name": ".NET Core Attach",
            "type": "coreclr",
            "request": "attach",
            "processId": "${command:pickProcess}"
        }
    ]
}
```
   And I still have unavailable `$exception` variable of course for this case for `PAUSED ON EXCEPTION`:
<img width="990" alt="2019-11-06_142058_1" src="https://user-images.githubusercontent.com/39411904/68303224-7711f400-00b4-11ea-99cc-6821f33f90ac.png">

2) I still find this behavior as unexpected and not intuitive for me by the next reasons:

    - I expect that we should see exception Window, because by default `Unhandled` exception filter always enabled.
    - I expect absolutely the same behavior for raising exception from Main thread and for any other user threads. Let's look. If I put one line in BUG1.cs in Main() than I have `Unhandled` exception with empty exception filter. I can't feel any differences for this two cases (BUG1.cs vs this line):

```
public static void Main() {
        throw new System.DivideByZeroException("TestSystemException0");
...
```
  
As you can see, we have wonderful exception Window for line above:
<img width="990" alt="2019-11-06_143308_2" src="https://user-images.githubusercontent.com/39411904/68304105-28655980-00b6-11ea-877d-3f89be9c18e3.png">

3) And also I think that we have illogical behavior for enabled `all` exception filter for BUG1.cs:
    - For this case I see two exception events. Why? Its really correct? (this looks like as a semantic dependence from a logical flag)

4) Please look at the next example. This is test on Exceptions which raised at the the same time from many threads. [BUG10.cs.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3814753/BUG10.cs.txt). For this test `All` enabled exception filter are needed. This example with undetermined problem on evaluation `$exception`. For reproducing are needed to add `$exception` to `Watch` Window and after each continue step (F5) to click (open/close) randomly by members in `$exception`. Next picture after about 5 exception events:
<img width="991" alt="2019-11-06_150430_3" src="https://user-images.githubusercontent.com/39411904/68305335-8eeb7700-00b8-11ea-9eea-3fe7d740b81b.png">

**Comment by gregg-miskelly** [comment]
> And I still have unavailable $exception variable of course for this case for PAUSED ON EXCEPTION

Yes, VS Code will not automaticially select a frame without symbols. So if you want to look at `$exception` you need to manually select it from the call stack window.

> I still find this behavior as unexpected and not intuitive for me

Yes, VS Code's behavior with functions that don't have associated source isn't wonderful. We have had a feature on the back log to try and improve this for a while, but it isn't something we have gotten to. https://github.com/OmniSharp/omnisharp-vscode/issues/1791

> I expect absolutely the same behavior for raising exception from Main thread and for any other user threads

The runtime doesn't act the same between the two scenarios. In the case of the main thread, the app just crashes without first catching the exception. Thus you will see 'Main' on the stack. In the case of the background thread - the framework catches and rethrows the exception. Thus by the time we crash there is no user code left on the stack.

> And also I think that we have illogical behavior for enabled all

I am assuming you mean you have checked 'All Exceptions' and NOT 'User-Unhandled Exceptions' (like in your finial screen capture). If so, I believe you are correct here that vsdbg's behavior is wrong. It is stopping for both thrown and 'user-unhandled', but really 'user-unhandled' should be controlled via the other setting.

> $exception is unavailable

Yes, `$exception` will only be available if there is an active frame, and if that active frame is a 'real' frame (not just an annotated frame like '[External Code]'). Its not the easiest user scenario to improve, so I would generally recommend leaving user-unhandled checked to avoid the problem (at least in most cases). At least until https://github.com/OmniSharp/omnisharp-vscode/issues/1791 is fixed.

**Comment by a-zh** [comment]
Gregg, thanks. These were all my current questions.

**Comment by gregg-miskelly** [comment]
Your are welcome. BTW, I am making two fixes in response to your questions --

1. As mentioned above, I fixed it so checking 'All Exceptions' doesn't implicitly turn on user-unhandled exceptions too
2. I fixed it so that, in VS Code, if we get a stack trace request on a thread where we are stopped on an exception, and that thread has only external code, we will show the call stack with external code filtering disabled. This is not enough to provide a good experience in VS Code. We really need https://github.com/OmniSharp/omnisharp-vscode/issues/1791 for that. But it is less bad in that one can at least evaluate $exception after manually selecting a frame.

**Comment by a-zh** [comment]
Thanks a lot. I would like to check these fixes. Let me know when updates will be ready.

**Comment by a-zh** [comment]
@gregg-miskelly I found a one another question for multi thread exceptions in `BUG10.cs.txt` test. After first `unhandled` exception event in some thread debugger has invalid user process state and cannot be allow to continue any other threads? Is it right that for multi thread applications we can see only one `unhandled` exception in any ways?

**Comment by gregg-miskelly** [comment]
Fixed in 1.21.
